### PR TITLE
Fix enum deadlock

### PR DESCRIPTION
**Pull request by jeremycarroll**
# Fixing a Deadlock and Documentation

I ended up finding it difficult to move from my initial fix, to a testable deployable change,
that I had run with the unit tests in Coursera. Hence this PR now covers two very different issues:

- fixing a race condition that sometimes creates a deadlock
- fixing the documentation, and some build scripts etc. to make it easier for a new person to make small changes and successfully deploy them

## Actual Intended Change

This PR addresses the deadlock seen in a [Coursera internal diff](https://phabricator.dkandu.me/D110457).
This deadlock appears to be an instance of the lazy value initialization order problem described in this [SIP](https://docs.scala-lang.org/sips/improved-lazy-val-initialization.html)

There are two new pieces of code:

1. The diff adds a test case exhibiting the deadlock

1. The laziness of the lazy field `properties` of `ScalaEnumTemplate` is re-implemented explicitly with no locking, so that multiple threads can initialize the value concurrently, with the last thread winning. This is semantically safe since all threads compute the same value. It resolves the deadlock because one of the locks no longer exists.

Prior to the implementation change, the `raceWithNameWithName` test usually (but not always) fails with the two critical threads having the following state - showing the deadlock between a class initialization lock for a companion object and the lock for the lazy property.
```
"B@1806" prio=5 tid=0x18 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	 blocks A@1805
	  at sun.misc.Unsafe.ensureClassInitialized(Unsafe.java:-1)
	  at sun.reflect.UnsafeFieldAccessorFactory.newFieldAccessor(UnsafeFieldAccessorFactory.java:43)
	  at sun.reflect.ReflectionFactory.newFieldAccessor(ReflectionFactory.java:156)
	  at java.lang.reflect.Field.acquireFieldAccessor(Field.java:1088)
	  at java.lang.reflect.Field.getFieldAccessor(Field.java:1069)
	  at java.lang.reflect.Field.get(Field.java:393)
	  at org.coursera.common.reflect.CompanionReflector$.findCompanionInstanceOfCompanionClass(CompanionReflector.scala:68)
	  at org.coursera.common.collection.Enum$$anonfun$findSymbols$1.applyOrElse(Enum.scala:79)
	  at org.coursera.common.collection.Enum$$anonfun$findSymbols$1.applyOrElse(Enum.scala:77)
	  at scala.PartialFunction.$anonfun$runWith$1$adapted(PartialFunction.scala:145)
	  at scala.PartialFunction$$Lambda$29.147022238.apply(Unknown Source:-1)
	  at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
	  at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
	  at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
	  at scala.collection.TraversableLike.collect(TraversableLike.scala:274)
	  at scala.collection.TraversableLike.collect$(TraversableLike.scala:272)
	  at scala.collection.mutable.ArrayOps$ofRef.collect(ArrayOps.scala:198)
	  at org.coursera.common.collection.Enum.findSymbols(Enum.scala:77)
	  at org.coursera.common.collection.Enum.findSymbols$(Enum.scala:63)
	  at org.coursera.courier.templates.ScalaEnumTemplate.findSymbols(ScalaEnumTemplate.scala:28)
	  at org.coursera.common.collection.Enum.symbols(Enum.scala:61)
	  at org.coursera.common.collection.Enum.symbols$(Enum.scala:61)
	  at org.coursera.courier.templates.ScalaEnumTemplate.symbols$lzycompute(ScalaEnumTemplate.scala:28)
	  - locked <0x716> (a org.coursera.courier.templates.EnumTypeB$)
	  at org.coursera.courier.templates.ScalaEnumTemplate.symbols(ScalaEnumTemplate.scala:28)
	  at org.coursera.courier.templates.EnumTypeB$.withName(EnumTemplateRaceTest.scala:199)
	  at org.coursera.courier.templates.TestClassB.withName(EnumTemplateRaceTest.scala:32)
	  at sun.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-1)
	  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	  at java.lang.reflect.Method.invoke(Method.java:498)
	  at org.coursera.courier.templates.EnumTemplateRaceTest$TestSetup.withName(EnumTemplateRaceTest.scala:95)
	  at org.coursera.courier.templates.EnumTemplateRaceTest.$anonfun$raceWithNameWithName$2(EnumTemplateRaceTest.scala:143)
	  at org.coursera.courier.templates.EnumTemplateRaceTest$$Lambda$35.1859383896.apply$mcV$sp(Unknown Source:-1)
	  at org.coursera.courier.templates.EnumTemplateRaceTest$TestSetup$$anon$3.run(EnumTemplateRaceTest.scala:108)

"A@1805" prio=5 tid=0x17 nid=NA waiting for monitor entry
  java.lang.Thread.State: BLOCKED
	 waiting for B@1806 to release lock on <0x716> (a org.coursera.courier.templates.EnumTypeB$)
	  at org.coursera.courier.templates.ScalaEnumTemplate.properties$lzycompute(ScalaEnumTemplate.scala:44)
	  at org.coursera.courier.templates.ScalaEnumTemplate.properties(ScalaEnumTemplate.scala:44)
	  at org.coursera.courier.templates.ScalaEnumTemplate.properties(ScalaEnumTemplate.scala:51)
	  at org.coursera.courier.templates.EnumTypeB$VALUE3$.<init>(EnumTemplateRaceTest.scala:188)
	  at org.coursera.courier.templates.EnumTypeB$VALUE3$.<clinit>(EnumTemplateRaceTest.scala:-1)
	  at org.coursera.courier.templates.TestClassB.value(EnumTemplateRaceTest.scala:31)
	  at sun.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-1)
	  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	  at java.lang.reflect.Method.invoke(Method.java:498)
	  at org.coursera.courier.templates.EnumTemplateRaceTest$TestSetup.value(EnumTemplateRaceTest.scala:94)
	  at org.coursera.courier.templates.EnumTemplateRaceTest.$anonfun$raceWithNameWithName$1(EnumTemplateRaceTest.scala:143)
	  at org.coursera.courier.templates.EnumTemplateRaceTest$$Lambda$34.640808588.apply$mcV$sp(Unknown Source:-1)
	  at org.coursera.courier.templates.EnumTemplateRaceTest$TestSetup$$anon$2.run(EnumTemplateRaceTest.scala:102)

```

The test case, CompanionRaceTest, is a version of this problem with no reference to any Coursera code. It uses two lazy fields and a companion class. If setting the `WORKAROUND` field to `false`, then the test fails, while it is `true` the test case uses the same workaround as is implemented in this diff. This is hence a variant of the Scala [lazy value problem](https://docs.scala-lang.org/sips/improved-lazy-val-initialization.html).

## Documentation and Build Changes

As I approach my real task, it became clear that the documentation was incomplete and messy.

- [CONTRIBUTING.md](https://github.com/coursera/courier/blob/749674fa7ee33804ad11b6c8ecb560f455cb4c22/CONTRIBUTING.md) had many errors, and very little that actually worked as written
- there was a second Coursera internal document that was crucial for getting started
- there were many omissions, which I had to discover by myself
- there were additional plugins (maven and gradle) which the code suggested should have been deployed, but in fact have not been updated for a long time, yet some scripts malfunctioned because of the gradle plugin.

I have made the following high-level changes to address these problems:

1. I consolidated both the Coursera internal document and the old CONTRIBUTING.md  into a new set of developer documents in a docs folder in this PR
1. I fixed the travis tests, which were broken
1. I moved the cross version part of the travis tests from .travis.yml to the definition of fulltest, to enable the simple documentation of _This project also uses the convention of prefixing commands with `full`  as a custom cross-version command_
1. I ensured the test suite from the courier-runtime project is actually run.
1. I improved the cross-build, so as to simplify the publishing steps
1. I made it clear that `courier-maven-plugin` and `courier-gradle-plugin` are not supported, removing the gradle support from some of the scripts (but clarifying and improving the documentation in the gradle subproject itself)
1. I ensured that all the commands suggested in the developer documentation actually work
1. I updated both the sonatype and bintray plugins to the final version before they ceased supporting sbt 0.13
1. I added @amory-coursera and @dguo-coursera to the list of developers since the currently published list has no active maintainers, which feels like an error.
1. I ensured that we can build and publish SNAPSHOT builds for integration testing in Coursera (both the sbt-plugin and the other subprojects; and even the gradle plugin)

### Testing Actual Change

The new unit test and running the tests on the [companion PR](https://github.com/webedx-spark/infra-services/pull/1722) over Coursera's services: many times, none of which deadlocked.

### Testing Other Changes

1. [Travis is building correctly](https://travis-ci.org/github/coursera/courier)
1. I read through the new documentation, and deleted my copy of the project and started from scratch (not the credentials part)

### CODE REVIEW

File: scala/runtime/src/test/scala/org/coursera/courier/templates/EnumTemplateRaceTest.scala [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    testSetup.value(0)
+    testSetup.withName(0)
+    assert(testSetup.race(testSetup.value(1), testSetup.withName(1)))
+  }
+
+}
+
+object EnumTemplateRaceTest {
+  // This test is for a heisenbug. If it recurs, please increase this value.
+  val NUMBER_OF_RUNS = 10
```
Comment (jeremycarroll) [inline scala/runtime/src/test/scala/org/coursera/courier/templates/EnumTemplateRaceTest.scala]: I will lower this to 1

File: scala/runtime/src/test/scala/org/coursera/courier/templates/EnumTemplateRaceTest.scala [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  case object VALUE2 extends EnumTypeA("VALUE2", properties("VALUE2"))
+  val SCHEMA = DataTemplateUtil.parseSchema(
+    """
+      { "type": "enum",
+        "name": "EnumTypeA",
+        "namespace": "org.coursera.courier.templates",
+        "symbols": ["VALUE1", "VALUE2"]
+      }""").asInstanceOf[EnumDataSchema]
+  override def withName(s: String): EnumTypeA = {
+    symbols.find(_.toString == s).get
```
Comment (jeremycarroll) [inline scala/runtime/src/test/scala/org/coursera/courier/templates/EnumTemplateRaceTest.scala:179]: Note this depends on [findSymbols](https://github.com/coursera/courscala/blob/master/courscala/src/main/scala/org/coursera/common/collection/Enum.scala#L63) which may be the code that might actually need 'fixing' ...

**Comment by mkowaliszyn-coursera** [comment]
For posterity, this affects program service quite frequently and probably every other service: https://docs.google.com/document/d/1L35IIDNzFUlMn8kjbU4o9-dBJGgjmFlH3NKMABvxXKo/edit#

**Comment by jeremycarroll** [comment]
image
image
image
image
image
image
image
image

**Comment by jeremycarroll** [comment]
image

**Comment by jeremycarroll** [comment]
image
image

**Comment by jeremycarroll** [comment]
Note my comments with images are a way to get the images into the documentation

File: docs/ARTIFACTS.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+The sbt-plugin is published to [bintray](https://bintray.com/coursera/sbt-plugins/courier-sbt-plugin)
+
+## Sonatype
+
+### Releases
+
+The Sonatype UI is not easy to use, for example, search does not
+seem to work well.
+
+Here is a path, through it.
```
Comment (amory-coursera) [inline docs/ARTIFACTS.md]:
```suggestion
Here is a path through it.
```

File: project/Sonatype.scala [hunk 4]
```diff
@@ -54,21 +54,29 @@ object Sonatype {
...
-            <id>Josh Newman</id>
-            <name>josh</name>
+            <id>josh</id>
+            <name>Josh Newman</name>
           </developer>
           <developer>
             <id>saeta</id>
             <name>Brennan Saeta</name>
           </developer>
+          <developer>
```
Comment (amory-coursera) [inline project/Sonatype.scala:72]: At this point, I think you've become a top committer @jeremycarroll . Consider adding yourself here as well?

File: scala/runtime/src/main/scala/org/coursera/courier/templates/ScalaEnumTemplate.scala [hunk 5]
```diff
@@ -41,11 +41,24 @@ abstract class ScalaEnumTemplate[T <: ScalaEnumTemplateSymbol]
   /**
    * Schema properties defined on this enum, if any.
    */
-  lazy val properties: Option[DataMap] = {
-    Option(SCHEMA.getProperties.get(symbolProperties)).collect {
-      case enumProps: DataMap => enumProps
+  def properties: Option[DataMap] =
+  // Implementation note: using a lazy field can result in deadlock.
```
Comment (amory-coursera) [inline scala/runtime/src/main/scala/org/coursera/courier/templates/ScalaEnumTemplate.scala]: [edit]
I see you've explained deadlock can occur when initializing objects with the class loader. A comment that references the test, where you've documented in detail, would be helpful. 

[original comment below]

I don't doubt that this is true given the effort you've put in to land this, but it's not obvious to me why this particular lazy val is problematic. I read [this article](https://blog.codecentric.de/en/2016/02/lazy-vals-scala-look-hood/) to gather some context, and the deadlock scenario they describe is due to an initialization chain that crosses synchronization boundaries. Is there a similar cause here? 

Reading the code, initialization seems pretty innocent, doesn't smell like a 'dangerous' lazy usage. Hoping to understand better, as if I'm ignorant here, we've probably got lots of code floating around where laziness introduces the possibility of a similar failure mode.

File: scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    assert(testSetup.race(testSetup.value(), testSetup.withName()))
+  }
+
+  /**
+    * This is [[CompanionRaceTest.raceToDeadlock]] modified to avoid the race
+    * by running one of the halves first, which initializes all the classes,
+    * we then change the instance in `testSetup` and run the race, which does
+    * not deadlock - showing that the class initialization lock is involved in
+    * the deadlock.
+    */
```
Comment (amory-coursera) [inline scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala:101]: This is very clever. I can follow along with some effort, but I don't think I would have come up with the series of refinements that trace this down to class initialization.

File: scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    testSetup.withName()
+    testSetup.changeInstance()
+    assert(testSetup.race(testSetup.value(), testSetup.withName()))
+  }
+
+}
+
+abstract class TestCompanion(properties: Option[String]) {}
+
+object TestCompanion {
```
Comment (amory-coursera) [inline scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala:114]: This appears to be a frozen state in the evolution from the original Courier enum source to your proposed solution. Were you to write a blog post about this, I'd find it a little easier to follow with a simpler object, e.g. using lazy constant values.

File: scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+  // Set this to false to exhibit the deadlock.
+  private val WORKAROUND = true
+  lazy val symbols: Set[TestCompanion] = {
+    Thread.sleep(10)
```
Comment (amory-coursera) [inline scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala]: Ha! At what time threshold do you start to see this behavior reliably? I assume it's rare without the sleep...

**Review by amory-coursera — APPROVED** [review]
Two really excellent pieces of work:

1. Detective work around class loader / lazy val deadlock is really interesting.
2. Updating contributor docs and fixing build scripts will make it much easier for the next person who wants to ship a patch.

Thank you for all the work, and the significant perseverance to get this ready to land.

File: scala/runtime/src/main/scala/org/coursera/courier/templates/ScalaEnumTemplate.scala (hunk 5 above)
Comment (jeremycarroll) [reply scala/runtime/src/main/scala/org/coursera/courier/templates/ScalaEnumTemplate.scala]: There are two lazy vals and a class lock, the other lazy val is in courscala project I think ... The article you linked to is good: _we can easily get into the same situation by using other concurrency and synchronization primitives_ Our other synchronization primitive is in class loading, somewhere.

File: scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala (hunk 7 above)
Comment (jeremycarroll) [reply scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala:114]: I chatted very briefly with the scala people, and they were well aware of the issue. I don't think I have much to add to the article you link. It is true, that I started with my original example and removed everything from coursera and got to this

File: scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala (hunk 8 above)
Comment (jeremycarroll) [reply scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala]: I don't recall why there is this sleep.

File: project/Sonatype.scala (hunk 4 above)
Comment (jeremycarroll) [reply project/Sonatype.scala:72]: I will do that

**Comment by jeremycarroll** [comment]
@amory-coursera one final question is the version number. I started by viewing this as a bug fix, taking the version to 2.1.8; we could say that the developer documentation is a new feature, taking it to 2.2.0; or we could even say that the (belated) honesty about gradle and maven not being supported is not backwardly compatible taking it to 3.0.0.

**Comment by jeremycarroll** [comment]
As per slack conversation, 


- [x] I am publishing 2.1.8 from this branch as is; 
- [x] I will make changes as requested from @amory-coursera, 
- [x] bump version to 3.0.0-SNAPSHOT
- [x] publish snapshot
- [x] test snapshot internally
- [x] bump version to 3.0.0
- [x] publish
- [x] update internal usage to 3.0.0
- [x] bump version to 3.0.1-SNAPSHOT

Ask someone with write access to:

- merge PR

- grant me write access

File: scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala (hunk 8 above)
Comment (jeremycarroll) [reply scala/runtime/src/test/scala/org/coursera/courier/templates/CompanionRaceTest.scala]: testing again today the sleep seems unnecessary; I have removed it

**Comment by jeremycarroll** [comment]
image

File: maven-plugin/pom.xml [hunk 9]
```diff
@@ -113,7 +113,7 @@
           <plugin>
             <groupId>org.apache.maven.plugins</groupId>
             <artifactId>maven-source-plugin</artifactId>
-            <version>3.0.0</version>
+            <version>3.0.1-SNAPSHOT</version>
```
Comment (kenwang-coursera) [inline maven-plugin/pom.xml:116]: Hey just curious if this line change was intentional? Trying to push out a PR for a 3.0.2-SNAPSHOT and not sure if this is supposed to be updated or not.

Comment (jeremycarroll) [reply maven-plugin/pom.xml:116]: that looks like an error ... (much later he says)

**Pull request merged**